src/consensus/pbft/libbyz/test/test_ledger_replay.cpp

```diff
@@ -166,124 +198,68 @@ TEST_CASE("Test Ledger Replay")
-          default:
-            INFO(
-              "We should only encounter ledger_append or ledger_truncate "
-              "messages");
-            REQUIRE(false);
-        }
-        ++num_msgs;
-      });
-    REQUIRE(num_msgs == (total_requests - 1));
-
-    // check that nothing gets executed out of order
-    INFO("replay entries 5 till 9 should fail");
-    auto first = initial_ledger.begin() + positions.at(4);
-    auto last = initial_ledger.end();
-    std::vector<uint8_t> vec_5_9(first, last);
-    CHECK(!replica->apply_ledger_data(vec_5_9));
-
-    INFO("replay entries 1 till 3");
-    first = initial_ledger.begin();
-    last = initial_ledger.begin() + positions.at(3);
-    std::vector<uint8_t> vec_1_3(first, last);
-    CHECK(replica->apply_ledger_data(vec_1_3));
-
-    INFO("replay entries 2 till 3 should fail");
-    first = initial_ledger.begin() + positions.at(1);
-    last = initial_ledger.begin() + positions.at(3);
-    std::vector<uint8_t> vec_2_3(first, last);
-    CHECK(!replica->apply_ledger_data(vec_2_3));
-
-    INFO("replay the rest of the pre-prepares in batches of 4");
-    for (size_t i = 3; i < total_requests - 1; i += 4)
+    replica->activate_pbft_local_hooks();
+    // ledgerenclave work
+    std::vector<std::vector<uint8_t>> entries;
+    while (true)
     {
-      auto until = i + 4;
-      first = initial_ledger.begin() + positions.at(i);
-      if (until >= positions.size())
+      auto ret = write_consensus->replay_data();
+      if (!ret.second)
       {
-        last = initial_ledger.end();
+        break;
+      }
+      // TODO: when deserialise will be called by pbft, in that place pbft will
+      // have to also append the write set to the ledger
+      entries.emplace_back(ret.first);
+    }
+    // apply out of order first
+    REQUIRE(
+      store->deserialise(entries.back()) == kv::DeserialiseSuccess::FAILED);
+
+    ccf::Store::Tx tx;
+    auto req_view = tx.get_view(pbft_requests);
+    auto req = req_view->get(0);
+    REQUIRE(!req.has_value());
+
+    auto pp_view = tx.get_view(pbft_pre_prepares);
+    auto pp = pp_view->get(0);
+    REQUIRE(!pp.has_value());
+
+    REQUIRE(entries.size() > 0);
+
+    Seqno seqno = 1;
+    size_t iterations = 0;
+    // apply all of the data in order
+    for (const auto& entry : entries)
+    {
+      REQUIRE(store->deserialise(entry) == kv::DeserialiseSuccess::PASS);
+      ccf::Store::Tx tx;
+      if (iterations % 2)
+      {
+        // odd entries are pre prepares
+        auto pp_view = tx.get_view(pbft_pre_prepares);
+        auto pp = pp_view->get(0);
+        REQUIRE(pp.has_value());
+        REQUIRE(pp.value().seqno == seqno);
+        seqno++;
       }
       else
       {
-        last = initial_ledger.begin() + positions.at(until);
+        // even entries are requests
+        auto req_view = tx.get_view(pbft_requests);
+        auto req = req_view->get(0);
+        REQUIRE(req.has_value());
+        REQUIRE(req.value().raw.size() > 0);
```

Comment:
I am not sure these two lines test much? Since the requests are always written at index `0`, this test only asserts that a request was written _at some point_ but not necessarily in the latest iteration.

Reply:
This is mostly here to make sure that the replicated data got deserialised and the derived data didn't. We could potentially add another check on the actual request to make sure that everything is being deserialised correctly. The main check is done further down where we see if `last_executed == total_requests -1` but I'll see if I can get more useful checking here

Reply:
added extra check in exec_command to make sure requests are getting executed in order